File: TargetedDosColisionsAndRNGAngle/testbed/metrics/fairness.py
"""Jain's fairness index (P4, AC-5) and polarization index (plan-8, OQ8-1)."""
from __future__ import annotations

from typing import Sequence


def jains_index(values: Sequence[float]) -> float:
    """`(sum(x))**2 / (n * sum(x**2))`. Returns `1.0` for the all-zero
    vector (no traffic = trivially fair, plan-4 Design) so an idle poll
    never divides by zero."""
    n = len(values)
    if n == 0:
        return 1.0
    total = sum(values)
    if total == 0:
        return 1.0
    sum_sq = sum(v * v for v in values)
    return (total**2) / (n * sum_sq)


def polarization_index(values: Sequence[float]) -> float:
    """`max(values) / mean(values)` (OQ8-1) -- scale-free (epic s3.4). `1.0`
    for the empty or all-zero vector, mirroring `jains_index`'s guard."""
    n = len(values)
    if n == 0:
        return 1.0
    total = sum(values)
    if total == 0:
        return 1.0
    mean = total / n
    return max(values) / mean

File: TargetedDosColisionsAndRNGAngle/web/lib/fairness.ts
// TS mirror of testbed/metrics/fairness.py -- Jain's index + polarization
// index (plan-8 OQ8-1), both scale-free ratios computed from real bucket counts.

export function jainsIndex(values: number[]): number {
  const n = values.length;
  if (n === 0) return 1;
  const total = values.reduce((a, b) => a + b, 0);
  if (total === 0) return 1;
  const sumSquares = values.reduce((a, b) => a + b * b, 0);
  return (total * total) / (n * sumSquares);
}

export function polarizationIndex(values: number[]): number {
  const n = values.length;
  if (n === 0) return 1;
  const total = values.reduce((a, b) => a + b, 0);
  if (total === 0) return 1;
  const mean = total / n;
  return Math.max(...values) / mean;
}
